mxnet/profile.go

```diff
@@ -38,44 +38,78 @@ type Profile struct {
+        ProfileMemoryDisable = ProfileMode(0)
+        ProfileApiDisable = ProfileMode(0)
+        ProfileContiguousDumpDisable = ProfileMode(0)
+        ProfileAllEnable  = ProfileMode(1)
+        ProfileSymbolicOperatorsEnable = ProfileMode(1)
+        ProfileImperativeOperatorsEnable = ProfileMode(1)
+        ProfileMemoryEnable = ProfileMode(1)
+        ProfileApiEnable = ProfileMode(1)
+        ProfileContiguousDumpEnable = ProfileMode(1)
+        ProfileDumpPeriod = ProfileMode(1)
 )
 
 // go binding for MXSetProfilerConfig()
-// param mode kOnlySymbolic: 0, kAllOperator: 1
-// param filename output filename
-func NewProfile(mode ProfileMode, tmpDir string) (*Profile, error) {
-	if tmpDir == "" {
-		tmpDir = filepath.Join(config.App.TempDir, "mxnet", "profile")
-	}
-	if !com.IsDir(tmpDir) {
-		os.MkdirAll(tmpDir, os.FileMode(0755))
-	}
-	filename, err := tempFile(tmpDir, "profile-", ".json")
-	if err != nil {
-		return nil, errors.Errorf("cannot create temporary file in %v", tmpDir)
-	}
-
-	cs := C.CString(filename)
-	defer C.free(unsafe.Pointer(cs))
-	success, err := C.MXSetProfilerConfig(C.int(mode), cs)
-	if err != nil {
-		return nil, err
-	}
-	if success != 0 {
-		return nil, GetLastError()
-	}
+// param profile_options map of profiling options
+// param tmpDir output filepath
+// CHANGE: modified interface to support new profiling options
+// TODO: make conversion neater if possible
+// TODO: check functionality
+func NewProfile(profileOptions map[string]ProfileMode, tmpDir string) (*Profile, error) {
+        if tmpDir == "" {
+                tmpDir = filepath.Join(config.App.TempDir, "mxnet", "profile")
+        }
+        if !com.IsDir(tmpDir) {
+                os.MkdirAll(tmpDir, os.FileMode(0755))
+        }
+        filename, err := tempFile(tmpDir, "profile-", ".json")
+        if err != nil {
+                return nil, errors.Errorf("cannot create temporary file in %v", tmpDir)
+        }
+
+        cs := C.CString(filename)
+        keys := [8]string{"filename", "profile_all", "profile_symbolic", "profile_imperative", "profile_memory","profile_api", "contiguous_dump", "dump_period"}
+
+        // convert go data structures into c data structures
+        ckeys := C.malloc(C.size_t(8) * C.size_t(unsafe.Sizeof(uintptr(0))))
+        a := (*[1<<30 - 1]*C.char)(ckeys)
+        cvals := C.malloc(C.size_t(8) * C.size_t(unsafe.Sizeof(uintptr(0))))
+        b := (*[1<<30 - 1]*C.char)(cvals)
+        a[0] = C.CString("filename")
+        b[0] = cs
+        for i:= 1 ; i < 8 ; i++ {
+                a[i] = C.CString(keys[i])
+                b[i] = C.CString(string(profileOptions[keys[i]]))
+        }
+
+        success, err := C.MXSetProfilerConfig(C.int(len(keys)), (**C.char)(ckeys), (**C.char)(cvals))
+        if err != nil {
+                return nil, err
+        }
+        if success != 0 {
+                return nil, GetLastError()
+        }
+
+        // free C pointers
+        C.free(unsafe.Pointer(cs))
```

Comment:
prefer to use defer over this, because you may return an error and this will cause a memory leak